firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
+                            let Some((child_addr, pos)) = children_iter.next() else {
+                                // We visited all this node's descendants. Go back to its parent.
+                                continue;
+                            };
+
+                            // Visit the child next.
+                            let child = merkle
+                                .get_node(child_addr)
+                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                            let partial_path = match child.inner() {
+                                NodeType::Branch(branch) => branch.path.iter().copied(),
+                                NodeType::Leaf(leaf) => leaf.path.iter().copied(),
+                                NodeType::Extension(_) => {
+                                    panic!("extension nodes shouldn't exist")
+                                }
+                            };
+
+                            let child_key: Box<[u8]> = key
+                                .iter()
+                                .copied()
+                                .chain(once(pos))
+                                .chain(partial_path)
+                                .collect();
+
+                            // There may be more children of this node to visit.
+                            iter_stack.push(iter_node);
+
+                            let child = merkle
+                                .get_node(child_addr)
+                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                            iter_stack.push(IterationNode::Unvisited {
+                                key: child_key,
+                                node: child,
+                            });
+                            return self.poll_next(_cx);
+                        }
+                    }
+                }
+                Poll::Ready(None)
+            }
+        }
+    }
+}
 
-                        Ok((std::mem::take(key), value))
-                    });
+// TODO comment
+fn get_iterator_intial_state<'a, S: ShaleStore<Node> + Send + Sync, T>(
+    merkle: &'a Merkle<S, T>,
+    root_node: DiskAddress,
+    key: &[u8],
+) -> Result<MerkleNodeStreamState<'a>, api::Error> {
+    // Invariant: `node`'s key is a prefix of `key`.
+    let mut node = merkle
+        .get_node(root_node)
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                    visited_node_path.push((found_node, 0));
+    // Invariant: This is `node`'s address at the start
+    // of each loop iteration.
+    let mut node_addr = root_node;
 
-                    self.key_state = IteratorState::Iterating {
-                        check_child_nibble,
-                        visited_node_path,
-                    };
+    // Invariant: [matched_key_nibbles] is the key of `node` at the start
+    // of each loop iteration.
+    let mut matched_key_nibbles = vec![];
```

Comment:
Can you use `Vec::with_capacity` here to minimize reallocations?

Reply:
🤔... maybe that wouldn't be good

Reply:
🤷 Honestly you would know better than me lol

Reply:
I was thinking that `with_capacity` would be better if you know the max size, but you're cloning the `matched_key_nibbles` as it grows, so you don't want a bunch of oversized vectors that will never grow. That would cause excess memory usage in an environment where memory is prized.